packages/block-library/src/image/edit.js

```diff
@@ -345,6 +349,80 @@ export function ImageEdit( {
 		className: classes,
 	} );
 
+	async function onHTMLDrop( HTML ) {
+		const blocks = pasteHandler( { HTML, mode: 'BLOCKS' } );
+
+		if ( ! blocks || blocks.length !== 1 ) {
+			return;
+		}
+
+		const [ block ] = blocks;
+
+		if ( ! block || block.name !== 'core/image' ) {
+			return;
+		}
+
+		// Optimistic update to replace the current image.
+		setAttributes( {
+			url: undefined,
+			alt: undefined,
+			id: undefined,
+			title: undefined,
+			caption: undefined,
+			...block.attributes,
+		} );
+
+		// Media item already exists in the library.
+		if ( !! block.attributes.id ) {
+			selectBlock( clientId );
+			return;
+		}
+
+		// Media item doesn't exist in the library, upload it so that it
+		// can be served within the same origin.
+		try {
+			// Media item does not exist in library, so try to upload it.
+			// Fist fetch the image data. This may fail if the image host
+			// doesn't allow CORS with the domain.
+			// If this happens, we insert the image block using the external
+			// URL and let the user know about the possible implications.
+			const response = await window.fetch( block.attributes.url );
+			const blob = await response.blob();
+
+			mediaUpload( {
+				filesList: [ blob ],
+				additionalData: {
+					title: block.attributes.title,
+					alt_text: block.attributes.alt,
+					caption: block.attributes.caption,
+				},
+				onFileChange( [ img ] ) {
+					if ( isBlobURL( img.url ) ) {
+						return;
+					}
+
+					__unstableMarkNextChangeAsNotPersistent();
+					setAttributes( {
+						id: img.id,
+						url: img.url,
+					} );
+					createSuccessNotice(
+						__( 'Image uploaded to the media library.' ),
+						{ type: 'snackbar' }
+					);
+				},
+				allowedTypes: ALLOWED_MEDIA_TYPES,
+				onError( message ) {
+					createErrorNotice( message, { type: 'snackbar' } );
+				},
+			} );
+		} catch ( err ) {
+			// TODO: Ignore cross-origin image errors for now.
```

Comment:
I think it should be possible to also show a confirmation modal here as shown in the media tab. Maybe we could just move the code from the media tab to here. I can look into this later.

A better solution, however, is to allow uploading external "url"s in the `/wp/v2/media` API, where the backend won't have limitations of the CORS errors. I'm not sure why it wasn't done this way?